rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,119 @@
+
+A different approach, as implemented by the Microsoft [Dev Tunnels SSH Library](https://github.com/microsoft/dev-tunnels-ssh) as the [session-reconnect@microsoft.com](https://github.com/microsoft/dev-tunnels-ssh/blob/main/ProtocolExtensions.md#extension-session-reconnection) extension, involves resuming the bidirectional stream of SSH packets after re-authenticating to the same server. This, unfortunately, would require the SSH library we use ([`golang.org/x/crypto/ssh`](https://golang.org/x/crypto/ssh)) to support this, meaning that we would need to fork the library, implement this nonstandard and fairly obscure extension, and then maintain our fork in the hopes that it would eventually get picked up upstream. In addition, this requires client support, meaning that we wouldn't be able to resume sessions opened via openssh or PuTTY (without some really nasty tricks like running a SSH server in `tsh proxy ssh`) - goal 3.
+
+This leads to the final approach, which is what we propose in this RFD: resuming the actual bytestream used by the SSH connection.
+
+Since we need some wire protocol anyway for the last leg of a connection to a direct dial node - we don't want resumption to be limited to tunnels, goal 4 - we might as well use the very same wire protocol directly from the client to the server. We initially thought of just relying on the tunneling protocol to let us carry the data stream as usual plus some needed metadata, but the way we currently do client-proxy and proxy-proxy connection forwarding (with frames sent over gRPC) doesn't actually allow for that without essentially employing the same multiplexing trickery that can be done end-to-end to begin with.
+
+Such protocol doesn't have to be SSH-specific, but at least some initial handshake must be, as we intend to support direct dial servers and we must not break compatibility with existing clients or servers - goal 0.
+
+Going over the goals again, we end up with this list:
+
+0. Do not break compatibility with existing servers or clients.
+
+1. Users of server access should not be able to notice that the Proxy service was restarted while they have a shell open, or Server Access shells should be able to survive a connection drop.
+
+1. Command execution and port forwarding should be able to survive a connection drop.
+
+1. OpenSSH compatibility via `tsh proxy ssh` should be able to survive a connection drop.
+
+1. Resilience against connection drops should include SSH servers running in "direct dial" mode, not just servers connected to a Teleport cluster via reverse tunnel.
```

Comment:
If a connection is terminated at the proxy (in proxy recording mode or when dealing with an openssh/agentless node) we can do resumption on the client-proxy side, but resuming a client-proxy connection is really hard as our current model doesn't let clients connect to a specific proxy, so you'd then have to throw in proxy peering to reach the specific proxy that has terminated the initial connection.

If proxy recording mode is enabled and if a Teleport node is in a leaf cluster we could benefit from doing resumption on the proxy-node connection (allowing the other proxies to restart, or allowing the connection to just accidentally drop) but that seems like a lot of effort for very little benefit, since the main benefit of proxy recording mode (recording OpenSSH sshd sessions) is now superseded by the dedicated openssh/agentless mode (and I don't think we want to build a connection resumption frontend for non-Teleport SSH servers, unless our connectivity model changes enough to make that plausible).

Reply:
Clarified "direct dial" mode in https://github.com/gravitational/teleport/pull/33439/commits/1caaf192031e73e082525d3ee8804106c4a6350c and detailed some possible extensions of connection resumption to proxy-terminated SSH connections (with their drawbacks) in https://github.com/gravitational/teleport/pull/33439/commits/2cf1b10548a304a7d767e1ff1249c249a4f3fd08